go/felix/intdataplane/floating_ip_mgr.go

```diff
@@ -0,0 +1,163 @@
+// table with DNAT and SNAT rules for the floating IPs associated with local workload endpoints.
+// The cali-fip-dnat chain is statically linked from cali-OUTPUT and cali-PREROUTING, and
+// cali-fip-snat from cali-POSTROUTING.
+type floatingIpManager struct {
+	ipVersion uint8
+
+	// Our dependencies.
+	natTable     iptablesTable
+	ruleRenderer rules.RuleRenderer
+
+	// Internal state.
+	activeDnatChains []*iptables.Chain
+	activeSnatChains []*iptables.Chain
+	natInfo          map[proto.WorkloadEndpointID][]*proto.NatInfo
+	dirtyNatInfo     bool
+}
+
+func newFloatingIpManager(
+	natTable iptablesTable,
+	ruleRenderer rules.RuleRenderer,
+	ipVersion uint8,
+) *floatingIpManager {
+	return &floatingIpManager{
+		natTable:     natTable,
+		ruleRenderer: ruleRenderer,
+		ipVersion:    ipVersion,
+
+		activeDnatChains: []*iptables.Chain{},
+		activeSnatChains: []*iptables.Chain{},
+		natInfo:          map[proto.WorkloadEndpointID][]*proto.NatInfo{},
+		dirtyNatInfo:     true,
+	}
+}
+
+func (m *floatingIpManager) OnUpdate(protoBufMsg interface{}) {
+	switch msg := protoBufMsg.(type) {
+	case *proto.WorkloadEndpointUpdate:
+		if m.ipVersion == 4 {
+			m.natInfo[*msg.Id] = msg.Endpoint.Ipv4Nat
+		} else {
+			m.natInfo[*msg.Id] = msg.Endpoint.Ipv6Nat
+		}
+		m.dirtyNatInfo = true
+	case *proto.WorkloadEndpointRemove:
+		delete(m.natInfo, *msg.Id)
+		m.dirtyNatInfo = true
+	}
+}
+
+func (m *floatingIpManager) CompleteDeferredWork() error {
+	if m.dirtyNatInfo {
+		// Collate required DNATs.
+		dnats := map[string]string{}
+		for _, natInfos := range m.natInfo {
+			for _, natInfo := range natInfos {
+				log.WithFields(log.Fields{
+					"ExtIp": natInfo.ExtIp,
+					"IntIp": natInfo.IntIp,
+				}).Debug("NAT mapping")
+
+				// We shouldn't ever have the same floating IP mapping to multiple
+				// workload IPs, but if we do we'll program the mapping to the
+				// alphabetically earlier one.
+				existingIntIp := dnats[natInfo.ExtIp]
+				if existingIntIp == "" || natInfo.IntIp < existingIntIp {
+					log.Debug("Wanted NAT mapping")
+					dnats[natInfo.ExtIp] = natInfo.IntIp
+				}
+			}
+		}
+		// Collate required SNATs.
+		snats := map[string]string{}
+		for extIp, intIp := range dnats {
+			log.WithFields(log.Fields{
+				"ExtIp": extIp,
+				"IntIp": intIp,
+			}).Debug("Reverse mapping")
+
+			// For the reverse SNATs, if multiple floating IPs map to the same workload
+			// IP, use the alphabetically earliest floating IP.
+			existingExtIp := snats[intIp]
+			if existingExtIp == "" || extIp < existingExtIp {
+				log.Debug("Wanted reverse mapping")
+				snats[intIp] = extIp
+			}
+		}
+		// Render chains for those NATs.
+		dnatChains := m.ruleRenderer.DNATsToIptablesChains(dnats)
+		snatChains := m.ruleRenderer.SNATsToIptablesChains(snats)
+		// Update iptables if they have changed.
+		if !reflect.DeepEqual(m.activeDnatChains, dnatChains) {
+			m.natTable.UpdateChains(dnatChains)
```

Comment:
Might be best to remove the old chains here first, in case we add a tree or something to the NAT chains later.